dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java

```diff
@@ -474,24 +468,11 @@ public void applicationStarting(ApplicationInfo appInfo) throws StateChangeExcep
 
             ClassLoader appCL = getRealAppClassLoader(application);
             if (appCL != null) {
-                newCL = classLoadingSRRef.getServiceWithException().createThreadContextClassLoader(appCL);
-                appTccls.put(appInfo.getName(), newCL);
-                oldCl = CDIUtils.getAndSetLoader(newCL);
+                ClassLoader newCL = classLoadingSRRef.getServiceWithException().createThreadContextClassLoader(appCL);
+                application.setTCCL(newCL);
             }
 
-            //Because weld fires observes in all modules when endInitialization() is called
-            //We can only set the jndi context once. This is sufficent for the java:app namespace
-            //but not for the java module namespace.
-
-            //Origonally I tried to setup JNDI so only application metadata was on the thread but
-            //that didn't work so I use give classic utils one of the module archives.
-
-            if (application.getModuleArchives().size() > 0 &&
-                application.getApplicationMetaData() != null) {
-                CDIArchive archive = application.getModuleArchives().iterator().next();
-                beginContext(archive);
-                setContext = true;
-            }
+            contextBeginnerEnder.extractComponentMetaData(application).extractTCCL(application).beginContext();
```

Comment:
It is in a try-with-resources. `contextBeginnerEnder` is created within the try statement, and its contents are filled in here after `application` has been created.

Reply:
The pattern with try-with-resources is that you open a resource (i.e. apply the context) in the try() and it gets closed (i.e. the context is removed) at the end of the block. Applying the context in the middle like this is horrible.

Splitting `ContextBeginnerEnder` into two objects as I originally suggested would prevent this because the `AutoClosable` object would only be created when `beginContext()` is called.

In addition, if an exception is thrown before `beginContext()` is called, `close()` will also throw an exception.